### DISCUSSION TITLE
Gradient shape oversampling

### DISCUSSION THREAD

**Discussion by m-a-x-i-m-z** [pulseq/pulseq#74] [Ideas]
Hi, 

I've just committed an updated demo spiral sequence, which now demonstrates the use of the new concept that I would like to test to see if it is useful. My goal is to improve the accuracy of the k-space calculation routines for non-Cartesian sequences that rely on arbitrary / shaped gradients. Another point is that I'd like to get rid of the annoying warning that the restoration of the missing shape points has failed (very much familiar to the majority of "spiral people" here). 

The idea is that for shaped gradients we now have an option to provide both the samples on the gradient raster centers (the ones that are used by the Siemens hardware) ans also on the gradient raster edges (which will be discarded by the interpreter once this feature is fully implemented). The good news is that we now can skip the sample restoration restoration code, which was producing strange results now and then. Additionally,  we eliminate potential instabilities / inconsistencies in the shape interpolation. Siemens implicitly uses no-interpolation and centered sample placement, in Pulseq we were assuming linear interpolation. These differences, (especially together with the failing sample restoration) may lead (or better to say have sometimes led) to undesired results. 

The plan is to include this feature into the 1.5.0 format update / release. 

Please check the code and share your thoughts here.

**Comment by zhengliuX** [comment]
Hi,   

I want to make sure that I undersrand it correctly. Now, in v1.4.x, for extended trap gradient, the gradient waveform is defined on the raster edge, for arbitrary gradient, the waveform is defined on the raster center. But in v1.5.x, both extended trap or arbitrary grad waveforms are defined on the raster center. Right?   

The reason I want to make sure that I understand it correcly, is because in UIH, we define all the gradient waveform on the raster edge, even for arbitrary gradients. For extended gradient waveforms, I don't need to modify the waveform, because all the waveforms are defined on the raster edge. but for arbitrary gradients, I need to re-calculate the value on the raster edge based on the waveform defined in the raster center using linear interpolation, which is also guaranteed by our spectrometer developers.   

Best, 
Zheng

**Reply by m-a-x-i-m-z** [reply]
> Now, in v1.4.x, for extended trap gradient, the gradient waveform is defined on the raster edge, for arbitrary gradient, the waveform is defined on the raster center.

Yes, this is correct. And it stays so in v1.5.0 as the normal use case.

> But in v1.5.x, both extended trap or arbitrary grad waveforms are defined on the raster center. Right?

Not exactly. In the proposed v1.5.0 format the gradient waveform may be _optionally_ stored on a raster that is _denser_ by a factor of 2 in comparison to what is defined by the gradRastertime. This means that the waveform will in this case have samples *both* on raster edges and raster centers. This means for the UI interpreter that you will not have to interpolate the samples on the edges but instead just take the samples provided (and ignore those in the centers). Please keep in mind that the first and the last sample (which are now also stored in the file) are not part of the waveform vector.

**Comment by m-a-x-i-m-z** [comment]
The current proposed gradient oversampling format (dev git branch) does the following: it stores the time vector in the gradient with a stepping of half gradient raster time. For the example spiral the gradients look like: 

```
[GRADIENTS]
4      -842987           -0      -842764 6 7 900
5       842986            0      19404.7 8 7 900
7      -842764      -842764           -0 9 10 0
8      19404.7      19404.7            0 9 10 0
```

Here, the first two gradients (#4 and #5) are the spiral readouts (regularly sampled gradients with oversampling), whereas gradients #7 and #8 are extended trapezoids (ramp down/spoiler gradients). This can be seen from the time shapes: 

```
[SHAPES]
...
shape_id 7
num_samples 7167
0.5
0.5
7165
...
shape_id 10
num_samples 2
0
187
```

Shape #7 is a compressed shape, containing 7167 regularly-sampled linearly increasing values with the step of 0.5. Shape #10 is an uncompressed shape containing exactly two samples (as needed to describe the ramp-down). 

If the gradients #4 and #5 were not over-sampled, then, instead of referring to the time shape #7 they would have a time shape ID of 0. 

Current Matlab code in the _dev_ branch can write and read the file in the above convention and also correctly interpret them. It detects the extended gradients by irregular time stepping (other that 1 or 0.5 gradient raster units).

However, the C++ code relies heavily on the just-in-time object decoding, so at the time, where we try to detect the type of the gradient the time shape has not been accessed (and of course not decoded). It therefore make it difficult to distinguish between the **extended trapezoids** and **arbitrary gradients with oversampling**. 

One idea how we could solve this is to use a negative _time_shape_ID_ as an indicator for oversampling. It would be rather easy to implement, however, the question comes, which numbers we should use. One possibility would be to have the negative value to indicate the oversampling factor, in which case only -2 will be supported at present. 

Any thoughts?

**Reply by zhengliuX** [reply]
If os being 2 is general, I suppose a negative value is enough. If we need a more flexible design, could we add a property, like "os", stored in the Pulseq file, to tell us the oversampling value.
